xblock/test/test_mixins.py

```diff
@@ -473,3 +478,109 @@ def test_datetime_serialization(self, value, expected_attributes):
+    """
+    class TestBuddyXBlock(XBlock, CompletableXBlockMixin):
+        """
+        Simple XBlock extending CompletableXBlockMixin.
+        """
+
+    class TestIllegalCustomCompletionAttrXBlock(XBlock, CompletableXBlockMixin):
+        """
+        XBlock extending CompletableXBlockMixin using illegal `has_custom_completion` attribute.
+        """
+        has_custom_completion = False
+
+    class TestIllegalCompletionMethodAttrXBlock(XBlock, CompletableXBlockMixin):
+        """
+        XBlock extending CompletableXBlockMixin using illegal `completion_method` attribute.
+        """
+        completion_method = "something_else"
+
+    def _make_block(self, runtime=None, block_type=None):
+        """
+        Creates a test block.
+        """
+        block_type = block_type if block_type else self.TestBuddyXBlock
+        runtime = runtime if runtime else mock.Mock(spec=Runtime)
+        scope_ids = ScopeIds("user_id", "test_buddy", "def_id", "usage_id")
+        return block_type(runtime=runtime, field_data=DictFieldData({}), scope_ids=scope_ids)
+
+    def test_has_custom_completion_property(self):
+        """
+        Test `has_custom_completion` property is set by mixin.
+        """
+        block = self._make_block()
+        self.assertTrue(block.has_custom_completion)
+        self.assertTrue(getattr(block, 'has_custom_completion', False))
+
+    def test_completion_method_property(self):
+        """
+        Test `completion_method` property is set by mixin.
+        """
+        block = self._make_block()
+        self.assertEqual(block.completion_method, 'completable')
+        self.assertEqual(getattr(block, 'completion_method', ""), 'completable')
+
+    @given(strategies.floats())
+    def test_emit_completion_illegal_custom_completion(self, any_completion):
+        """
+        Test `emit_completion` raises exception when called on a XBlock with illegal `has_custom_completion` value.
+        """
+        runtime_mock = mock.Mock(spec=Runtime)
+        illegal_custom_completion_block = self._make_block(runtime_mock, self.TestIllegalCustomCompletionAttrXBlock)
+        with self.assertRaises(AttributeError):
+            illegal_custom_completion_block.emit_completion(any_completion)
+
+    @given(strategies.floats())
+    def test_emit_completion_completion_method(self, any_completion):
+        """
+        Test `emit_completion` raises exception when called on a XBlock with illegal `completion_method` value.
+        """
+        runtime_mock = mock.Mock(spec=Runtime)
+        illegal_completion_method_block = self._make_block(runtime_mock, self.TestIllegalCompletionMethodAttrXBlock)
+        with self.assertRaises(AttributeError):
+            illegal_completion_method_block.emit_completion(any_completion)
+
+    @given(strategies.floats(min_value=0.0, max_value=1.0))
+    @example(1.0)
+    @example(0.0)
+    def test_emit_completion_emits_event(self, valid_completion_ratio):
```

Comment:
Reminder: when we decide on the proper terms, search the entire PR for "ratio" and replace with the "percent"-like terms.  Thanks.